python/pyarrow/serialization.pxi

```diff
@@ -43,93 +43,98 @@ class DeserializationException(Exception):
-        serialize objects of the class in a particular way.
-    custom_deserializer : callable
-        This argument is optional, but can be provided to
-        deserialize objects of the class in a particular way.
-    """
-    type_to_type_id[type] = type_id
-    whitelisted_types[type_id] = type
-    if pickle:
-        types_to_pickle.add(type_id)
-    if custom_serializer is not None:
-        custom_serializers[type_id] = custom_serializer
-        custom_deserializers[type_id] = custom_deserializer
-
-
-def _serialization_callback(obj):
-    if type(obj) not in type_to_type_id:
-        raise SerializationException("pyarrow does not know how to "
-                                     "serialize objects of type {}."
-                                     .format(type(obj)),
-                                     obj)
-    type_id = type_to_type_id[type(obj)]
-    if type_id in types_to_pickle:
-        serialized_obj = {"data": pickle.dumps(obj), "pickle": True}
-    elif type_id in custom_serializers:
-        serialized_obj = {"data": custom_serializers[type_id](obj)}
-    else:
-        if is_named_tuple(type(obj)):
-            serialized_obj = {}
-            serialized_obj["_pa_getnewargs_"] = obj.__getnewargs__()
-        elif hasattr(obj, "__dict__"):
-            serialized_obj = obj.__dict__
-        else:
-            raise SerializationException("We do not know how to serialize "
-                                         "the object '{}'".format(obj), obj)
-    return dict(serialized_obj, **{"_pytype_": type_id})
-
-
-def _deserialization_callback(serialized_obj):
-    type_id = serialized_obj["_pytype_"]
-
-    if "pickle" in serialized_obj:
-        # The object was pickled, so unpickle it.
-        obj = pickle.loads(serialized_obj["data"])
-    else:
-        assert type_id not in types_to_pickle
-        if type_id not in whitelisted_types:
-            raise "error"
-        type = whitelisted_types[type_id]
-        if type_id in custom_deserializers:
-            obj = custom_deserializers[type_id](serialized_obj["data"])
+cdef class SerializationContext:
+    cdef:
+        object type_to_type_id
+        object whitelisted_types
+        object types_to_pickle
+        object custom_serializers
+        object custom_deserializers
+
+    def __init__(self):
+        # Types with special serialization handlers
+        self.type_to_type_id = dict()
+        self.whitelisted_types = dict()
+        self.types_to_pickle = set()
+        self.custom_serializers = dict()
+        self.custom_deserializers = dict()
+
+    def register_type(self, type_, type_id, pickle=False,
+                      custom_serializer=None, custom_deserializer=None):
+        """Add type to the list of types we can serialize.
+
+        Parameters
+        ----------
+        type_ : type
+            The type that we can serialize.
+        type_id : bytes
+            A string of bytes used to identify the type.
+        pickle : bool
```

Comment:
Sort of unrelated to this PR, but should we be providing a way to set a different pickler, e.g., `cloudpickle` instead of `pickle`?

Reply:
Alternatively, maybe later we can just replace the whole pickle case with a custom_serializer/custom_deserializer that just uses `pickle` or `cloudpickle`.

Reply:
It already tries to use cloudpickle if it is installed, see
https://github.com/apache/arrow/blob/c0beb879e3c4863a10b0d09aa981e8f53776458a/python/pyarrow/compat.py#L133

For now I'm trying not to make SerializationContext too generic because we don't really know what is required, and is very easy to write a custom one.